development/gcp/cloud-functions/rotateserviceaccount/rotateserviceaccount.go

```diff
@@ -0,0 +1,125 @@
+	ProjectID        string `json:"project_id"`
+	PrivatekayID     string `json:"private_key_id"`
+	PrivateKey       string `json:"private_key"`
+	ClientEmail      string `json:"client_email"`
+	ClientID         string `json:"client_id"`
+	AuthURL          string `json:"auth_uri"`
+	TokenURI         string `json:"token_uri"`
+	AuthProviderCert string `json:"auth_provider_x509_cert_url"`
+	ClientCert       string `json:"client_x509_cert_url"`
+}
+
+func init() {
+	var err error
+	ctx := context.Background()
+
+	projectID, err = metadata.ProjectID()
+	if err != nil {
+		panic("failed to retrieve GCP Project ID, error: " + err.Error())
+	}
+
+	secretManagerService, err = secretmanager.NewService(ctx)
+	if err != nil {
+		panic("failed creating Secret Manager client, error: " + err.Error())
+	}
+
+	serviceAccountService, err = iam.NewService(ctx)
+	if err != nil {
+		panic("failed creating IAM client, error: " + err.Error())
+	}
+}
+
+// RotateServiceAccount adds new secret version in Secret Manager on pubsub rotate message
+func RotateServiceAccount(ctx context.Context, m pubsub.MessagePayload) error {
+	var err error
+	var secretRotateMessage pubsub.SecretRotateMessage
+	var secretData ServiceAccountJSON
+
+	// Create logger to use google cloud functions structured logging
+	logger := cloudfunctions.NewLogger()
+	// Set component for log entries to identify all messages for this function.
+	logger.WithComponent("kyma.prow.cloud-function.RotateServiceAccount")
+	// Set trace value for log entries to identify messages from one function call.
+	logger.GenerateTraceValue(projectID, "RotateServiceAccount")
+
+	if m.Attributes["eventType"] != "SECRET_ROTATE" {
+		logger.LogDebug("Unsupported event type: %s, quitting", m.Attributes["eventType"])
+		return nil
+	}
+
+	err = json.Unmarshal(m.Data, &secretRotateMessage)
+	if err != nil {
+		logger.LogCritical("failed to unmarshal message data field, error: %s", err.Error())
+	}
+
+	if secretRotateMessage.Labels["type"] != "service-account" {
+		logger.LogDebug("Unsupported secret type: %s, quitting", secretRotateMessage.Labels["type"])
+		return nil
+	}
+
+	//get latest secret version data
+	secretlatestVersionPath := secretRotateMessage.Name + "/versions/latest"
+	logger.LogInfo("Retrieving secret: %s", secretlatestVersionPath)
+	secretDataString, err := secretManagerService.GetSecretVersionData(secretlatestVersionPath)
+	if err != nil {
+		logger.LogCritical("failed to retrieve latest version of a secret %s, error: %s", secretRotateMessage.Name, err.Error())
+	}
+
+	logger.LogInfo("Trying to unmarshal secret: %s", secretRotateMessage.Name)
+	decodedSecretDataString, err := base64.StdEncoding.DecodeString(secretDataString)
+	if err != nil {
+		logger.LogCritical("Could not base64 decode secret %s", secretRotateMessage.Name)
+	}
+	err = json.Unmarshal([]byte(decodedSecretDataString), &secretData)
+	if err != nil {
+		logger.LogCritical("failed to unmarshal secret JSON field, error: %s", err.Error())
+	}
```

Comment:
Is /versions/latest a static value of path suffix to get latest version of secret? If yes, I would implement this part as method of secretManagerService and call it here for each secret name. I think this will be frequent usage of secretManagerService
secretManagerService.GetSecretLatestVersion(secretName string) (ServiceAccountJSON, error)

Reply:
I've added `GetLatestSecretVersion` and `GetLatestSecretVersionData` functions